### PR TITLE
Implement syscall operation

### PR DESCRIPTION
**Pull request by bobbinth**
This PR implements `SYSCALL` operation and the associated assembly instruction. Main changes are:

- Implement `Kernel` struct which is now included with every program.
- Update assembler to enable instantiation with a given kernel source and parsing programs in the context of the specified kernel.
- Update processor to execute programs in the context of a given kernel.

This PR does not include:
- Updates to the decoder required to enforce syscall semantics
- Updates to chiplets component to enforce verification of kernel calls
- Updates to the documentation

### CODE REVIEW

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @bobbinth! Looks good to me overall, but I think it would be good to add a bit more about compiling kernel procedures to the inline docs

File: core/src/program/blocks/call_block.rs [hunk 1]
```diff
@@ -24,7 +25,23 @@ impl Call {
...
+            hash,
+            fn_hash,
+            is_syscall: false,
+        }
+    }
+
+    /// Returns a new [Call] block instantiated with the specified function body hash and marked
+    /// as a kernel call.
+    pub fn new_syscall(fn_hash: Digest) -> Self {
+        // TODO: make hash computation different from regular call
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/program/blocks/call_block.rs:38]: I think this is a more general issue of how we compute hashes of nodes (we need to do something similar for differentiating between hashes of `JOIN` vs. `SPLIT` blocks). We'll address this as we migrate to a different version of Rescue hash function (probably not early than v0.4).

[the remark this answers is not in the document]
Comment (grjte) [reply core/src/program/blocks/call_block.rs:38]: Should that whole migration be captured as an issue then? (and just marked "on hold" for now)

File: assembly/src/lib.rs [hunk 2]
```diff
@@ -46,34 +46,72 @@ type ModuleMap = BTreeMap<String, ProcMap>;
 // ASSEMBLER
 // ================================================================================================
 
-/// TODO: add comments
+/// Miden Assembler which can be used to convert Miden assembly source code into program MAST (
+/// represented by the [Program] struct). The assembler can be instantiated in two ways:
+/// - Via the `with_kernel()` constructor. In this case, the specified kernel source is compiled
+///   into a set of kernel procedures during instantiation. The programs complied using such
```
Comment (grjte) [inline assembly/src/lib.rs:52]: nit: There's a typo. It says "complied" instead of "compiled"

**Review by grjte — APPROVED** [review]
Thanks @bobbinth ! I think the docs changes were helpful

**Pull request merged**